test/WebJobs.Script.Tests/Security/SecretManagerTests.cs

```diff
@@ -337,6 +340,55 @@ public async Task GetHostSecrets_UpdatesStaleSecrets()
             }
         }
 
+        [Fact]
+        public async Task GetFunctionSecretsAsync_SecretGenerationIsSerialized()
+        {
+            var mockValueConverterFactory = GetConverterFactoryMock(false, false);
+            var metricsLogger = new TestMetricsLogger();
+            var testRepository = new TestSecretsRepository(true);
+            string testFunctionName = $"TestFunction";
+
+            using (var secretManager = new SecretManager(testRepository, mockValueConverterFactory.Object, _logger, metricsLogger, _hostNameProvider, _startupContextProvider))
```

Comment:
This test reproduces the issue. I also ran this test using the actual Blob repository, and it resulted in the exception in the issue. However, since these unit tests shouldn't be communicating to storage, I've created a TestSecretsRepository that enforces serialized writes and catches the issue.